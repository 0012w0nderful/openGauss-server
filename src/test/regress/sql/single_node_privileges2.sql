-- test use trigger to promote authority
create user user01 password 'test-1234';
SET SESSION AUTHORIZATION user01 password 'test-1234';

create table user01.tbl01(a int);
-- function to promote authority
create or replace function user01.tg1() returns trigger as $$
declare
issysadmin boolean;
begin
  select rolsystemadmin into issysadmin from pg_roles where rolname = current_user;
  CASE issysadmin
  WHEN 1 THEN
    alter USER user01 with sysadmin; 
  else
  null;
  end case;
  return new;
end;
$$ language plpgsql security invoker; 

-- trigger for table
create trigger insert_tg1 
before insert on user01.tbl01
for each row
execute procedure user01.tg1();

grant all on user01.tbl01 to public;

-- try to trigger the trigger, should fail
insert into user01.tbl01 values(1);

RESET SESSION AUTHORIZATION;

-- use root user to trigger, should fail
insert into user01.tbl01 values(1);

-- cleanup
drop schema user01 cascade;
drop role user01;