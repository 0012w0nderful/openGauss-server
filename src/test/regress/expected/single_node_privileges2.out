-- test use trigger to promote authority
create user user01 password 'test-1234';
SET SESSION AUTHORIZATION user01 password 'test-1234';
create table user01.tbl01(a int);
-- function to promote authority
create or replace function user01.tg1() returns trigger as $$
declare
issysadmin boolean;
begin
  select rolsystemadmin into issysadmin from pg_roles where rolname = current_user;
  CASE issysadmin
  WHEN 1 THEN
    alter USER user01 with sysadmin; 
  else
  null;
  end case;
  return new;
end;
$$ language plpgsql security invoker; 
-- trigger for table
create trigger insert_tg1 
before insert on user01.tbl01
for each row
execute procedure user01.tg1();
grant all on user01.tbl01 to public;
-- try to trigger the trigger, should fail
insert into user01.tbl01 values(1);
ERROR:  permission denied for relation pg_roles
CONTEXT:  SQL statement "select rolsystemadmin                 from pg_roles where rolname = current_user"
PL/pgSQL function tg1() line 5 at SQL statement
RESET SESSION AUTHORIZATION;
-- use root user to trigger, should fail
insert into user01.tbl01 values(1);
ERROR:  permission denied for relation pg_roles
CONTEXT:  SQL statement "select rolsystemadmin                 from pg_roles where rolname = current_user"
PL/pgSQL function tg1() line 5 at SQL statement
-- cleanup
drop schema user01 cascade;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to table user01.tbl01
drop cascades to function user01.tg1()
drop role user01;
